### PR TITLE
DeformableDetrModel support fp16

### PR DESCRIPTION
**Pull request by DonggeunYu**
# What does this PR do?

This PR for DeformableDetrModel support fp16.

https://github.com/huggingface/transformers/issues/29011

## Who can review?

@amyeroberts

### CODE REVIEW

File: src/transformers/kernels/deformable_detr/cuda/ms_deform_attn_cuda.cu [hunk 1]
```diff
@@ -64,7 +64,7 @@ at::Tensor ms_deform_attn_cuda_forward(
     for (int n = 0; n < batch/im2col_step_; ++n)
     {
         auto columns = output_n.select(0, n);
-        AT_DISPATCH_FLOATING_TYPES(value.type(), "ms_deform_attn_forward_cuda", ([&] {
+        AT_DISPATCH_FLOATING_TYPES_AND_HALF(value.type(), "ms_deform_attn_forward_cuda", ([&] {
```
Comment (DonggeunYu) [inline src/transformers/kernels/deformable_detr/cuda/ms_deform_attn_cuda.cu:67]: Reference: https://github.com/nijkah/mmcv/blob/e242908146ccf0bf3900113022caab798cfa129f/mmcv/ops/csrc/pytorch/cuda/ms_deform_attn_cuda.cu#L329

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 2]
```diff
@@ -1540,14 +1540,14 @@ def unfreeze_backbone(self):
...
+    def get_valid_ratio(self, mask, dtype=torch.float32):
         """Get the valid ratio of all feature maps."""
 
         _, height, width = mask.shape
         valid_height = torch.sum(mask[:, :, 0], 1)
         valid_width = torch.sum(mask[:, 0, :], 1)
-        valid_ratio_heigth = valid_height.float() / height
-        valid_ratio_width = valid_width.float() / width
+        valid_ratio_heigth = valid_height.to(dtype) / height
+        valid_ratio_width = valid_width.to(dtype) / width
```
Comment (DonggeunYu) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1550]: Reference: https://github.com/huggingface/transformers/blob/de6029a0593d6ab73b4b0c6c71f5aa6e2520293f/src/transformers/models/oneformer/modeling_oneformer.py#L1355-L1364

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 3]
```diff
@@ -1540,14 +1540,14 @@ def unfreeze_backbone(self):
...
-    def get_valid_ratio(self, mask):
+    def get_valid_ratio(self, mask, dtype=torch.float32):
         """Get the valid ratio of all feature maps."""
 
         _, height, width = mask.shape
         valid_height = torch.sum(mask[:, :, 0], 1)
         valid_width = torch.sum(mask[:, 0, :], 1)
-        valid_ratio_heigth = valid_height.float() / height
-        valid_ratio_width = valid_width.float() / width
+        valid_ratio_heigth = valid_height.to(dtype) / height
```
Comment (amyeroberts) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:1549]:
```suggestion
        valid_ratio_height = valid_height.to(dtype) / height
```

**Review by amyeroberts — APPROVED** [review]
Very nice! 🔥 

Thanks for taking the time to link to all the relevant references, it really helps for a quick review. 

Only comment is a small spelling nit. I realise it was there before, but let's take the opportunity to fix it.

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py (hunk 3 above)
Comment (DonggeunYu) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py]: Thanks~

**Comment by DonggeunYu** [comment]
@amyeroberts
I took all of your feedback on board.

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding this great contribution to the library!

**Pull request merged**